servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -290,6 +281,15 @@ public void onNext(@Nullable final Collection<? extends ServiceDiscovererEvent<R
                 LOGGER.debug("{}: unexpectedly received null instead of events.", RoundRobinLoadBalancer.this);
                 return;
             }
+            // According to Reactive Streams Rule 1.8
+            // (https://github.com/reactive-streams/reactive-streams-jvm?tab=readme-ov-file#1.8) new events will
+            // stop eventually but not guaranteed to stop immediately after cancellation or could race with cancel.
+            // Therefore, we should check that this is the current Subscriber before processing new events.
+            if (healthCheckConfig != null && currentSubscriber != this) {
```

Comment:
Note that for the `RoundRobinLoadBalancer` there is still a possibility for a race if `ServiceDiscoverer` implementation doesn't follow the "use a single thread" recommendation: https://github.com/apple/servicetalk/pull/3002/files#diff-994a09bcaa716664549357bca8bfa8288ee0dd98b0280b139fd30b15cd69bd8eR30-R32

However, taking into account that:
1. we are close to switching to `DefaultLoadBalancer`;
2. the default DNS `ServiceDiscoverer` is single-threaded, and for other implementations it's a trivial fix;
3. possibility for that race is significantly reduced by the current check;

I don't see a reason to significantly overcomplicate this implementation unless someone has a simple suggestion.

Example of a race:
1. SD Thread#1 delivers events for old `EventSubscriber`, crosses this check and OS doesn't execute it further.
2. Caller Thread#2 triggers cancellation, resubscribes and flips the current subscriber.
3. SD Thread#3 delivers "state of the world" for a new `EventSubscriber`.
4. OS continues execution of Thread#1 and it corrupts the `usedHosts` state.

Reply:
I tend to agree with your assessment that we need not address it: this race has been here for a long time and we've survived until now so a little longer wont hurt. However, if you felt like fixing it I think it could be done with a lock protecting all `EventSubscriber.onNext` calls.